man/pam_systemd.8
'\" t
.TH "PAM_SYSTEMD" "8" "" "systemd 221" "pam_systemd"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
pam_systemd \- Register user sessions in the systemd login manager
.SH "SYNOPSIS"
.PP
pam_systemd\&.so
.SH "DESCRIPTION"
.PP
\fBpam_systemd\fR
registers user sessions with the systemd login manager
\fBsystemd-logind.service\fR(8), and hence the systemd control group hierarchy\&.
.PP
On login, this module ensures the following:
.sp
.RS 4
.ie n \{\
\h'-04' 1.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  1." 4.2
.\}
If it does not exist yet, the user runtime directory
/run/user/$USER
is created and its ownership changed to the user that is logging in\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 2.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  2." 4.2
.\}
The
\fI$XDG_SESSION_ID\fR
environment variable is initialized\&. If auditing is available and
\fBpam_loginuid\&.so\fR
was run before this module (which is highly recommended), the variable is initialized from the auditing session id (/proc/self/sessionid)\&. Otherwise, an independent session counter is used\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 3.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  3." 4.2
.\}
A new systemd scope unit is created for the session\&. If this is the first concurrent session of the user, an implicit slice below
user\&.slice
is automatically created and the scope placed into it\&. An instance of the system service
user@\&.service, which runs the systemd user manager instance, is started\&.
.RE
.PP
On logout, this module ensures the following:
.sp
.RS 4
.ie n \{\
\h'-04' 1.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  1." 4.2
.\}
If enabled in
\fBlogind.conf\fR(5), all processes of the session are terminated\&. If the last concurrent session of a user ends, the user\*(Aqs systemd instance will be terminated too, and so will the user\*(Aqs slice unit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 2.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  2." 4.2
.\}
If the last concurrent session of a user ends, the
\fI$XDG_RUNTIME_DIR\fR
directory and all its contents are removed, too\&.
.RE
.PP
If the system was not booted up with systemd as init system, this module does nothing and immediately returns
\fBPAM_SUCCESS\fR\&.
.SH "OPTIONS"
.PP
The following options are understood:
.PP
\fBclass=\fR
.RS 4
Takes a string argument which sets the session class\&. The XDG_SESSION_CLASS environmental variable takes precedence\&. One of
"user",
"greeter",
"lock\-screen"
or
"background"\&. See
\fBsd_session_get_class\fR(3)
for details about the session class\&.
.RE
.PP
\fBtype=\fR
.RS 4
Takes a string argument which sets the session type\&. The XDG_SESSION_TYPE environmental variable takes precedence\&. One of
"unspecified",
"tty",
"x11",
"wayland"
or
"mir"\&. See
\fBsd_session_get_type\fR(3)
for details about the session type\&.
.RE
.PP
\fBdebug\fR\fB[=]\fR
.RS 4
Takes an optional boolean argument\&. If yes or without the argument, the module will log debugging information as it operates\&.
.RE
.SH "MODULE TYPES PROVIDED"
.PP
Only
\fBsession\fR
is provided\&.
.SH "ENVIRONMENT"
.PP
The following environment variables are set for the processes of the user\*(Aqs session:
.PP
\fI$XDG_SESSION_ID\fR
.RS 4
A session identifier, suitable to be used in filenames\&. The string itself should be considered opaque, although often it is just the audit session ID as reported by
/proc/self/sessionid\&. Each ID will be assigned only once during machine uptime\&. It may hence be used to uniquely label files or other resources of this session\&.
.RE
.PP
\fI$XDG_RUNTIME_DIR\fR
.RS 4
Path to a user\-private user\-writable directory that is bound to the user login time on the machine\&. It is automatically created the first time a user logs in and removed on the user\*(Aqs final logout\&. If a user logs in twice at the same time, both sessions will see the same
\fI$XDG_RUNTIME_DIR\fR
and the same contents\&. If a user logs in once, then logs out again, and logs in again, the directory contents will have been lost in between, but applications should not rely on this behavior and must be able to deal with stale files\&. To store session\-private data in this directory, the user should include the value of
\fI$XDG_SESSION_ID\fR
in the filename\&. This directory shall be used for runtime file system objects such as
\fBAF_UNIX\fR
sockets, FIFOs, PID files and similar\&. It is guaranteed that this directory is local and offers the greatest possible file system feature set the operating system provides\&. For further details see the
\m[blue]\fBXDG Base Directory Specification\fR\m[]\&\s-2\u[1]\d\s+2\&.
.RE
.PP
The following environment variables are read by the module and may be used by the PAM service to pass metadata to the module:
.PP
\fI$XDG_SESSION_TYPE\fR
.RS 4
The session type\&. This may be used instead of
\fBsession=\fR
on the module parameter line, and is usually preferred\&.
.RE
.PP
\fI$XDG_SESSION_CLASS\fR
.RS 4
The session class\&. This may be used instead of
\fBclass=\fR
on the module parameter line, and is usually preferred\&.
.RE
.PP
\fI$XDG_SESSION_DESKTOP\fR
.RS 4
A single, short identifier string for the desktop environment\&. This may be used to indicate the session desktop used, where this applies and if this information is available\&. For example:
"GNOME", or
"KDE"\&. It is recommended to use the same identifiers and capitalization as for
\fI$XDG_CURRENT_DESKTOP\fR, as defined by the
\m[blue]\fBDesktop Entry Specification\fR\m[]\&\s-2\u[2]\d\s+2\&. (However, note that
\fI$XDG_SESSION_DESKTOP\fR
only takes a single item, and not a colon\-separated list like
\fI$XDG_CURRENT_DESKTOP\fR\&.) See
\fBsd_session_get_desktop\fR(3)
for more details\&.
.RE
.PP
\fI$XDG_SEAT\fR
.RS 4
The seat name the session shall be registered for, if any\&.
.RE
.PP
\fI$XDG_VTNR\fR
.RS 4
The VT number the session shall be registered for, if any\&. (Only applies to seats with a VT available, such as
"seat0")
.RE
.SH "EXAMPLE"
.sp
.if n \{\
.RS 4
.\}
.nf
#%PAM\-1\&.0
auth       required     pam_unix\&.so
auth       required     pam_nologin\&.so
account    required     pam_unix\&.so
password   required     pam_unix\&.so
session    required     pam_unix\&.so
session    required     pam_loginuid\&.so
session    required     pam_systemd\&.so
.fi
.if n \{\
.RE
.\}
.SH "SEE ALSO"
.PP
\fBsystemd\fR(1),
\fBsystemd-logind.service\fR(8),
\fBlogind.conf\fR(5),
\fBloginctl\fR(1),
\fBpam.conf\fR(5),
\fBpam.d\fR(5),
\fBpam\fR(8),
\fBpam_loginuid\fR(8),
\fBsystemd.scope\fR(5),
\fBsystemd.slice\fR(5),
\fBsystemd.service\fR(5)
.SH "NOTES"
.IP " 1." 4
XDG Base Directory Specification
.RS 4
\%http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html
.RE
.IP " 2." 4
Desktop Entry Specification
.RS 4
\%http://standards.freedesktop.org/desktop-entry-spec/latest/
.RE


debian/changelog
systemd (221-2) UNRELEASED; urgency=medium

  * Disable test-netlink during package build, fails on some buildds.
  * udev.postinst: Don't call addgroup with --quiet, so that if the "input"
    group already exists as a non-system group you get a sensible error
    message. Some broken tutorials forget the --system option.
    (Closes: #769948, LP: #1455956)
  * systemd.postinst: Drop the --quiet from the addgroup calls as well, same
    reason as above. (Closes: #762275)
  * udev: Drop doc dir symlinking. It has caused too much trouble and only
    marginally helps to avoid duplication. Such duplication should be dealt
    with at the distro, not package level.

 -- Martin Pitt <mpitt@debian.org>  Mon, 29 Jun 2015 07:42:50 +0200

systemd (221-1) unstable; urgency=medium

  * New upstream release 221:
    - Fix persistent storage links for Xen devices. (LP: #1467151)
    - Drop all backported patches and port the others to new upstream release.
    - debian/rules: Drop workarounds for broken 220 tarball, 221 is fine.

  [ Michael Biebl ]
  * initramfs hook: Stop installing 55-dm.rules, 64-md-raid.rules,
    60-persistent-storage-lvm.rules and 60-persistent-storage-dm.rules.
    The mdadm, lvm2 and dmsetup package provide their own udev hooks nowadays
    to make sure their udev rules files are installed into the initramfs.
    Having the copy rules at two places is confusing and makes debugging
    harder.
  * Make it possible to skip building udeb packages via
    DEB_BUILD_OPTIONS="noudeb". This allows quicker builds for local testing
    and is benefical for derivatives that don't use d-i.
  * Install API documentation for libudev and libsystemd in their respective
    packages. Both libraries use man pages now, so we need to be explicit
    about what is installed where.

  [ Martin Pitt ]
  * ifupdown-hotplug autopkgtest: Different cloud/desktop environments have
    different ways of including /etc/network/interfaces.d/, try to get along
    wit either and skip the test if interfaces.d/ does not get included at
    all.
  * Drop obsolete gtk-doc-tools build dependency, gtkdocize autoreconfig, and
    ./configure options.
  * libudev-dev.install: Drop gtk-doc files, not built by upstream any more
    and replaced with manpages.
  * libsystemd0.symbols: Add new symbols for this release.
  * debian/rules: Fix paths in manpages as we don't currently have a merged
    /usr in Debian but have most systemd things in /lib. This replaces the
    previous huge and maintenance-intense patch.
  * Drop Accept-mountall-specific-fstab-options.patch. Replaced with
    systemd.postinst migration code in Ubuntu.
  * Revert overly aggressive USB autosuspend udev rules change which broke
    various USB keyboards. (Closes: #789723)
  * Have rc-local.service output also go to the console. /etc/rc.local often
    contains status messages which users expect to see during boot.
    (LP: #1468102)
  * debian/rules: Install udev.NEWS into libudev1, to get along with Debian's
    udev -> libudev1 doc dir symlinking. (Closes: #790042)

 -- Martin Pitt <mpitt@debian.org>  Sun, 28 Jun 2015 12:05:36 +0200

systemd (220-7) unstable; urgency=medium

  [ Michael Biebl ]
  * Enable seccomp support on arm64 as well.
  * Replace the remainder of Fix-paths-in-man-pages.patch with an upstream
    provided patch.

  [ Martin Pitt ]
  * Switch to net.ifnames persistent network interfaces (on new
    installations/for new hardware), and deprecate the old
    75-persistent-net-generator.rules. See the ML discussion for details:
        https://lists.debian.org/debian-devel/2015/05/msg00170.html
        https://lists.debian.org/debian-devel/2015/06/msg00018.html
    - Drop Make-net.ifnames-opt-in-instead-of-opt-out.patch, to use
      net.ifnames by default.
    - Revert-udev-network-device-renaming-immediately-give.patch: Adjust
      patch comment.
    - Drop 75-persistent-net-generator.rules, write_net_rules helper and
      rule_generator.functions.
    - Adjust udev's README.Debian accordingly, and describe the migration.
      This needs to happen manually as there is no robust way of doing this
      automatically.
    - Add udev NEWS file for announcing this change and pointing to udev's
      README.
    - udev.postinst: Drop write_interfaces_rules().
    - udev.postinst: Disable net.ifnames on systems which did not support
      75-persistent-net-generator.rules (most importantly, virtualized guests)
      to avoid changing network interface names on upgrade.
    - LP: #1454254
  * fsckd-daemon-for-inter-fsckd-communication.patch: Add fsckd.c to
    POTFILES.in.
  * ifupdown-hotplug autopkgtest: Fix config name in interfaces.d/, it must
    not have a suffix in Debian. Also clean up the file after the test.
  * net.agent: When running under systemd, run everything in the foreground.
    This avoids killing the forked child in the middle of its operation under
    systemd when the parent exits.
  * Check during build that systemd and systemd-journald don't link against
    anything in /usr, to prevent bugs like #771652 and #788913 in the future.
  * Drop Skip-99-systemd.rules-when-not-running-systemd-as-in.patch. The rules
    mostly just attach tags systemd specific properties which are harmless
    under other init systems, and systemd-sysctl also works there.
  * 80-networking.rules: Only call agents for add|remove, as they don't handle
    other events.
  * Restore udev watches on block device changes. (Closes: #789060,
    LP: #1466081)


src/login/pam_systemd.c
/*-*- Mode: C; c-basic-offset: 8; indent-tabs-mode: nil -*-*/

/***
  This file is part of systemd.

  Copyright 2010 Lennart Poettering

  systemd is free software; you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation; either version 2.1 of the License, or
  (at your option) any later version.

  systemd is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with systemd; If not, see <http://www.gnu.org/licenses/>.
***/

#include <errno.h>
#include <fcntl.h>
#include <sys/file.h>
#include <pwd.h>
#include <endian.h>

#include <security/pam_modules.h>
#include <security/_pam_macros.h>
#include <security/pam_modutil.h>
#include <security/pam_ext.h>
#include <security/pam_misc.h>

#include "util.h"
#include "audit.h"
#include "macro.h"
#include "strv.h"
#include "bus-util.h"
#include "def.h"
#include "socket-util.h"
#include "fileio.h"
#include "bus-error.h"

static int parse_argv(
                pam_handle_t *handle,
                int argc, const char **argv,
                const char **class,
                const char **type,
                bool *debug) {

        unsigned i;

        assert(argc >= 0);
        assert(argc == 0 || argv);

        for (i = 0; i < (unsigned) argc; i++) {
                if (startswith(argv[i], "class=")) {
                        if (class)
                                *class = argv[i] + 6;

                } else if (startswith(argv[i], "type=")) {
                        if (type)
                                *type = argv[i] + 5;

                } else if (streq(argv[i], "debug")) {
                        if (debug)
                                *debug = true;

                } else if (startswith(argv[i], "debug=")) {
                        int k;

                        k = parse_boolean(argv[i] + 6);
                        if (k < 0)
                                pam_syslog(handle, LOG_WARNING, "Failed to parse debug= argument, ignoring.");
                        else if (debug)
                                *debug = k;

                } else
                        pam_syslog(handle, LOG_WARNING, "Unknown parameter '%s', ignoring", argv[i]);
        }

        return 0;
}

static int get_user_data(
                pam_handle_t *handle,
                const char **ret_username,
                struct passwd **ret_pw) {

        const char *username = NULL;
        struct passwd *pw = NULL;
        int r;

        assert(handle);
        assert(ret_username);
        assert(ret_pw);

        r = pam_get_user(handle, &username, NULL);
        if (r != PAM_SUCCESS) {
                pam_syslog(handle, LOG_ERR, "Failed to get user name.");
                return r;
        }

        if (isempty(username)) {
                pam_syslog(handle, LOG_ERR, "User name not valid.");
                return PAM_AUTH_ERR;
        }

        pw = pam_modutil_getpwnam(handle, username);
        if (!pw) {
                pam_syslog(handle, LOG_ERR, "Failed to get user data.");
                return PAM_USER_UNKNOWN;
        }

        *ret_pw = pw;
        *ret_username = username;

        return PAM_SUCCESS;
}

static int get_seat_from_display(const char *display, const char **seat, uint32_t *vtnr) {
        union sockaddr_union sa = {
                .un.sun_family = AF_UNIX,
        };
        _cleanup_free_ char *p = NULL, *tty = NULL;
        _cleanup_close_ int fd = -1;
        struct ucred ucred;
        int v, r;

        assert(display);
        assert(vtnr);

        /* We deduce the X11 socket from the display name, then use
         * SO_PEERCRED to determine the X11 server process, ask for
         * the controlling tty of that and if it's a VC then we know
         * the seat and the virtual terminal. Sounds ugly, is only
         * semi-ugly. */

        r = socket_from_display(display, &p);
        if (r < 0)
                return r;
        strncpy(sa.un.sun_path, p, sizeof(sa.un.sun_path)-1);

        fd = socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC, 0);
        if (fd < 0)
                return -errno;

        if (connect(fd, &sa.sa, offsetof(struct sockaddr_un, sun_path) + strlen(sa.un.sun_path)) < 0)
                return -errno;

        r = getpeercred(fd, &ucred);
        if (r < 0)
                return r;

        r = get_ctty(ucred.pid, NULL, &tty);
        if (r < 0)
                return r;

        v = vtnr_from_tty(tty);
        if (v < 0)
                return v;
        else if (v == 0)
                return -ENOENT;

        if (seat)
                *seat = "seat0";
        *vtnr = (uint32_t) v;

        return 0;
}

static int export_legacy_dbus_address(
                pam_handle_t *handle,
                uid_t uid,
                const char *runtime) {

#ifdef ENABLE_KDBUS
        _cleanup_free_ char *s = NULL;
        int r;

        /* skip export if kdbus is not active */
        if (access("/sys/fs/kdbus", F_OK) < 0)
                return PAM_SUCCESS;

        if (asprintf(&s, KERNEL_USER_BUS_ADDRESS_FMT ";" UNIX_USER_BUS_ADDRESS_FMT, uid, runtime) < 0) {
                pam_syslog(handle, LOG_ERR, "Failed to set bus variable.");
                return PAM_BUF_ERR;
        }

        r = pam_misc_setenv(handle, "DBUS_SESSION_BUS_ADDRESS", s, 0);
        if (r != PAM_SUCCESS) {
                pam_syslog(handle, LOG_ERR, "Failed to set bus variable.");
                return r;
        }
#endif
        return PAM_SUCCESS;
}

_public_ PAM_EXTERN int pam_sm_open_session(
                pam_handle_t *handle,
                int flags,
                int argc, const char **argv) {

        _cleanup_bus_error_free_ sd_bus_error error = SD_BUS_ERROR_NULL;
        _cleanup_bus_message_unref_ sd_bus_message *reply = NULL;
        const char
                *username, *id, *object_path, *runtime_path,
                *service = NULL,
                *tty = NULL, *display = NULL,
                *remote_user = NULL, *remote_host = NULL,
                *seat = NULL,
                *type = NULL, *class = NULL,
                *class_pam = NULL, *type_pam = NULL, *cvtnr = NULL, *desktop = NULL;
        _cleanup_bus_close_unref_ sd_bus *bus = NULL;
        int session_fd = -1, existing, r;
        bool debug = false, remote;
        struct passwd *pw;
        uint32_t vtnr = 0;
        uid_t original_uid;

        assert(handle);

        if (parse_argv(handle,
                       argc, argv,
                       &class_pam,
                       &type_pam,
                       &debug) < 0)
                return PAM_SESSION_ERR;

        if (debug)
                pam_syslog(handle, LOG_DEBUG, "pam-systemd initializing");

        r = get_user_data(handle, &username, &pw);
        if (r != PAM_SUCCESS) {
                pam_syslog(handle, LOG_ERR, "Failed to get user data.");
                return r;
        }

        /* Make sure we don't enter a loop by talking to
         * systemd-logind when it is actually waiting for the
         * background to finish start-up. If the service is
         * "systemd-user" we simply set XDG_RUNTIME_DIR and
         * leave. */

        pam_get_item(handle, PAM_SERVICE, (const void**) &service);
        if (streq_ptr(service, "systemd-user")) {
                _cleanup_free_ char *p = NULL, *rt = NULL;

                if (asprintf(&p, "/run/systemd/users/"UID_FMT, pw->pw_uid) < 0)
                        return PAM_BUF_ERR;

                r = parse_env_file(p, NEWLINE,
                                   "RUNTIME", &rt,
                                   NULL);
                if (r < 0 && r != -ENOENT)
                        return PAM_SESSION_ERR;

                if (rt)  {
                        r = pam_misc_setenv(handle, "XDG_RUNTIME_DIR", rt, 0);
                        if (r != PAM_SUCCESS) {
                                pam_syslog(handle, LOG_ERR, "Failed to set runtime dir.");
                                return r;
                        }

                        r = export_legacy_dbus_address(handle, pw->pw_uid, rt);
                        if (r != PAM_SUCCESS)
                                return r;
                }

                return PAM_SUCCESS;
        }

        /* Otherwise, we ask logind to create a session for us */

        pam_get_item(handle, PAM_XDISPLAY, (const void**) &display);
        pam_get_item(handle, PAM_TTY, (const void**) &tty);
        pam_get_item(handle, PAM_RUSER, (const void**) &remote_user);
        pam_get_item(handle, PAM_RHOST, (const void**) &remote_host);

        seat = pam_getenv(handle, "XDG_SEAT");
        if (isempty(seat))
                seat = getenv("XDG_SEAT");

        cvtnr = pam_getenv(handle, "XDG_VTNR");
        if (isempty(cvtnr))
                cvtnr = getenv("XDG_VTNR");

        type = pam_getenv(handle, "XDG_SESSION_TYPE");
        if (isempty(type))
                type = getenv("XDG_SESSION_TYPE");
        if (isempty(type))
                type = type_pam;

        class = pam_getenv(handle, "XDG_SESSION_CLASS");
        if (isempty(class))
                class = getenv("XDG_SESSION_CLASS");
        if (isempty(class))
                class = class_pam;

        desktop = pam_getenv(handle, "XDG_SESSION_DESKTOP");
        if (isempty(desktop))
                desktop = getenv("XDG_SESSION_DESKTOP");

        tty = strempty(tty);

        if (strchr(tty, ':')) {
                /* A tty with a colon is usually an X11 display,
                 * placed there to show up in utmp. We rearrange
                 * things and don't pretend that an X display was a
                 * tty. */

                if (isempty(display))
                        display = tty;
                tty = NULL;
        } else if (streq(tty, "cron")) {
                /* cron has been setting PAM_TTY to "cron" for a very
                 * long time and it probably shouldn't stop doing that
                 * for compatibility reasons. */
                type = "unspecified";
                class = "background";
                tty = NULL;
        } else if (streq(tty, "ssh")) {
                /* ssh has been setting PAM_TTY to "ssh" for a very
                 * long time and probably shouldn't stop doing that
                 * for compatibility reasons. */
                type ="tty";
                class = "user";
                tty = NULL;
        }

        /* If this fails vtnr will be 0, that's intended */
        if (!isempty(cvtnr))
                safe_atou32(cvtnr, &vtnr);

        if (!isempty(display) && !vtnr) {
                if (isempty(seat))
                        get_seat_from_display(display, &seat, &vtnr);
                else if (streq(seat, "seat0"))
                        get_seat_from_display(display, NULL, &vtnr);
        }

        if (seat && !streq(seat, "seat0") && vtnr != 0) {
                pam_syslog(handle, LOG_DEBUG, "Ignoring vtnr %"PRIu32" for %s which is not seat0", vtnr, seat);
                vtnr = 0;
        }

        if (isempty(type))
                type = !isempty(display) ? "x11" :
                           !isempty(tty) ? "tty" : "unspecified";

        if (isempty(class))
                class = streq(type, "unspecified") ? "background" : "user";

        remote = !isempty(remote_host) && !is_localhost(remote_host);

        /* Talk to logind over the message bus */

        r = sd_bus_open_system(&bus);
        if (r < 0) {
                pam_syslog(handle, LOG_ERR, "Failed to connect to system bus: %s", strerror(-r));
                return PAM_SESSION_ERR;
        }

        if (debug)
                pam_syslog(handle, LOG_DEBUG, "Asking logind to create session: "
                           "uid="UID_FMT" pid="PID_FMT" service=%s type=%s class=%s desktop=%s seat=%s vtnr=%"PRIu32" tty=%s display=%s remote=%s remote_user=%s remote_host=%s",
                           pw->pw_uid, getpid(),
                           strempty(service),
                           type, class, strempty(desktop),
                           strempty(seat), vtnr, strempty(tty), strempty(display),
                           yes_no(remote), strempty(remote_user), strempty(remote_host));

        r = sd_bus_call_method(bus,
                               "org.freedesktop.login1",
                               "/org/freedesktop/login1",
                               "org.freedesktop.login1.Manager",
                               "CreateSession",
                               &error,
                               &reply,
                               "uusssssussbssa(sv)",
                               (uint32_t) pw->pw_uid,
                               (uint32_t) getpid(),
                               service,
                               type,
                               class,
                               desktop,
                               seat,
                               vtnr,
                               tty,
                               display,
                               remote,
                               remote_user,
                               remote_host,
                               0);
        if (r < 0) {
                pam_syslog(handle, LOG_ERR, "Failed to create session: %s", bus_error_message(&error, r));
                return PAM_SYSTEM_ERR;
        }

        r = sd_bus_message_read(reply,
                                "soshusub",
                                &id,
                                &object_path,
                                &runtime_path,
                                &session_fd,
                                &original_uid,
                                &seat,
                                &vtnr,
                                &existing);
        if (r < 0) {
                pam_syslog(handle, LOG_ERR, "Failed to parse message: %s", strerror(-r));
                return PAM_SESSION_ERR;
        }

        if (debug)
                pam_syslog(handle, LOG_DEBUG, "Reply from logind: "
                           "id=%s object_path=%s runtime_path=%s session_fd=%d seat=%s vtnr=%u original_uid=%u",
                           id, object_path, runtime_path, session_fd, seat, vtnr, original_uid);

        r = pam_misc_setenv(handle, "XDG_SESSION_ID", id, 0);
        if (r != PAM_SUCCESS) {
                pam_syslog(handle, LOG_ERR, "Failed to set session id.");
                return r;
        }

        if (original_uid == pw->pw_uid) {
                /* Don't set $XDG_RUNTIME_DIR if the user we now
                 * authenticated for does not match the original user
                 * of the session. We do this in order not to result
                 * in privileged apps clobbering the runtime directory
                 * unnecessarily. */

                r = pam_misc_setenv(handle, "XDG_RUNTIME_DIR", runtime_path, 0);
                if (r != PAM_SUCCESS) {
                        pam_syslog(handle, LOG_ERR, "Failed to set runtime dir.");
                        return r;
                }

                r = export_legacy_dbus_address(handle, pw->pw_uid, runtime_path);
                if (r != PAM_SUCCESS)
                        return r;
        }

        if (!isempty(seat)) {
                r = pam_misc_setenv(handle, "XDG_SEAT", seat, 0);
                if (r != PAM_SUCCESS) {
                        pam_syslog(handle, LOG_ERR, "Failed to set seat.");
                        return r;
                }
        }

        if (vtnr > 0) {
                char buf[DECIMAL_STR_MAX(vtnr)];
                sprintf(buf, "%u", vtnr);

                r = pam_misc_setenv(handle, "XDG_VTNR", buf, 0);
                if (r != PAM_SUCCESS) {
                        pam_syslog(handle, LOG_ERR, "Failed to set virtual terminal number.");
                        return r;
                }
        }

        r = pam_set_data(handle, "systemd.existing", INT_TO_PTR(!!existing), NULL);
        if (r != PAM_SUCCESS) {
                pam_syslog(handle, LOG_ERR, "Failed to install existing flag.");
                return r;
        }

        if (session_fd >= 0) {
                session_fd = fcntl(session_fd, F_DUPFD_CLOEXEC, 3);
                if (session_fd < 0) {
                        pam_syslog(handle, LOG_ERR, "Failed to dup session fd: %m");
                        return PAM_SESSION_ERR;
                }

                r = pam_set_data(handle, "systemd.session-fd", INT_TO_PTR(session_fd+1), NULL);
                if (r != PAM_SUCCESS) {
                        pam_syslog(handle, LOG_ERR, "Failed to install session fd.");
                        safe_close(session_fd);
                        return r;
                }
        }

        return PAM_SUCCESS;
}

_public_ PAM_EXTERN int pam_sm_close_session(
                pam_handle_t *handle,
                int flags,
                int argc, const char **argv) {

        _cleanup_bus_error_free_ sd_bus_error error = SD_BUS_ERROR_NULL;
        _cleanup_bus_close_unref_ sd_bus *bus = NULL;
        const void *existing = NULL;
        const char *id;
        int r;

        assert(handle);

        /* Only release session if it wasn't pre-existing when we
         * tried to create it */
        pam_get_data(handle, "systemd.existing", &existing);

        id = pam_getenv(handle, "XDG_SESSION_ID");
        if (id && !existing) {

                /* Before we go and close the FIFO we need to tell
                 * logind that this is a clean session shutdown, so
                 * that it doesn't just go and slaughter us
                 * immediately after closing the fd */

                r = sd_bus_open_system(&bus);
                if (r < 0) {
                        pam_syslog(handle, LOG_ERR, "Failed to connect to system bus: %s", strerror(-r));
                        return PAM_SESSION_ERR;
                }

                r = sd_bus_call_method(bus,
                                       "org.freedesktop.login1",
                                       "/org/freedesktop/login1",
                                       "org.freedesktop.login1.Manager",
                                       "ReleaseSession",
                                       &error,
                                       NULL,
                                       "s",
                                       id);
                if (r < 0) {
                        pam_syslog(handle, LOG_ERR, "Failed to release session: %s", bus_error_message(&error, r));
                        return PAM_SESSION_ERR;
                }
        }

        /* Note that we are knowingly leaking the FIFO fd here. This
         * way, logind can watch us die. If we closed it here it would
         * not have any clue when that is completed. Given that one
         * cannot really have multiple PAM sessions open from the same
         * process this means we will leak one FD at max. */

        return PAM_SUCCESS;
}
